R/geom_phylopic.R

```diff
@@ -36,61 +36,129 @@ download_phylopic <- function(id, destdir = ".", ...) {
-    width <- getOption("phylopic_width")
-    if (is.null(width))
-        width <- 256
+    #width <- getOption("phylopic_width")
+    #if (is.null(width))
+    #    width <- 256
 
+    #basepath <- getOption('phylopic_dir')
+    #if (is.null(basepath)) {
+    #    basepath <- "http://phylopic.org/assets/images/submissions"
+    #}         
+
+    #url <- paste0(basepath, "/", id, ".", width, ".png")
+
+    #if (is.null(basepath))
+    #    url <- check_url(url)
+    #return(url)
     basepath <- getOption('phylopic_dir')
-    if (is.null(basepath)) {
-        basepath <- "http://phylopic.org/assets/images/submissions"
-    }         
+    if (is.null(basepath)){
+        basepath <- "https://images.phylopic.org/images/"
+    }
 
-    url <- paste0(basepath, "/", id, ".", width, ".png")
+    id <- .autocomplete_uid(x = id)
 
-    if (is.null(basepath))
+    url <- paste0(basepath, id, "/vector.svg")
+
+    if (is.null(basepath)){
         url <- check_url(url)
+    }
+
     return(url)
+    
 }
 
-phylopic_valid_id <- function(id) {
-    res <- vapply(id, phylopic_valid_id_item, character(1))
-    i <- which(res == "")
-    res[i] <- NA
-    return(res)
+# Universally unique identifier (uuid) of phylopic database 
+# is a 128-bit number. It has 32 alphanumeric characters in the
+# form of 8-4-4-4-12.
+.autocomplete_uid <- function(x){
+    x1 <- unlist(strsplit(x, split='-'))
+    flag1 <- length(x1) == 5
+    if (!flag1){
+        x <- phylopic_uid(name = x)$uid
+    }else{
+        flag2 <- all(nchar(x1) == c(8, 4, 4, 4, 12))
+        if (!flag2){
+            x <- phylopic_uid(name = x)$uid
+        }
+    }
+    return(x)
 }
 
-phylopic_valid_id_item <- function(id) {
-    url <- paste0("http://phylopic.org/api/a/name/", id,
-                  "/images?subtaxa=true&supertaxa=true&options=pngFiles+canoicalName+json")
-    res <- tryCatch(jsonlite::fromJSON(url)$result,
-                    error = function(e) return(NULL))
-    if (is.null(res)) return("")
-
-    if (length(res$same) > 0) {
-        taxa <- res$same
-    } else if (length(res$supertaxa) > 0) {
-        taxa <- res$supertaxa
-    } else if (length(res$subtaxa) > 0){
-        taxa <- res$subtaxa
-    } else if (length(res$other) > 0) {
-        taxa <- res$other
-    } else {
-        return("")
-    }
 
-    uid <- taxa$uid[1]
-    return(uid)
+##' obtaion suggestions for full names based on partial text name
+##'
+##'
+##' @title autocomplete_name
+##' @param name partial text name
+##' @param ... additional parameters
+##' @return scientific name
+##' @export
+autocomplete_name <- function(name, ...){
+    x <- lapply(name, .autocomplete_name_search)
+    x <- do.call('rbind', x)
+    return(x)
+}
+
+.autocomplete_name_search <- function(name, ...){
+    x <- gsub("[^a-zA-Z]+", "%20", tolower(name))
+    url <- paste0('https://api.phylopic.org/autocomplete?build=182&query=', x)
```

Comment:
Note that the build number can change, so you'll want to exclude that parameter (or fetch it first).